### DISCUSSION TITLE
Using sa.text() to preserve inline literals while leaving BindParameters as placeholders after compiling to raw SQL

### DISCUSSION THREAD

**Discussion by bskubi** [sqlalchemy/sqlalchemy#12206] [Usage Questions]
I'm new to SQLAlchemy and relatively new to relational databases. I'm working on a Python library that builds CRUD SQL statements using SQLAlchemy. This is for scientific computing and no untrustworthy values are being handled. The main database format I'm targeting is DuckDB, and while duckdb_engine is nice, I've found that executing SQLAlchemy statements using duckdb_engine directly is substantially slower than compiling the SQLAlchemy statements to raw SQL, then executing them with DuckDB's native connection and execution methods.

This poses a challenge, because compiling with literal_binds replaces any BindParameter values with NULL, while compiling without literal_binds replaces all literals with placeholders. I'd like my library functions to build queries that retain user-passed literals through the compilation process, while retaining the rest as placeholders.

The way I'm currently accomplishing this is by examining any user-passed parameters and converting them to a TextClause with sa.text() if they are strings or numbers. Here's a small example for illustrative purposes:

```
def build_insert(name: str = sa.bindparam("name"), num: int = sa.bindparam("num"), name_tbl):
    return (
        sa.insert(name_tbl)
        .values(
            name = sa.text("'{name}'") if isinstance(name, str) else name,
            num = sa.text("{num}") if isinstance(num, str) else num
        )
    )
```

This statement is designed to let the user get a fully-constructed SQLAlchemy statement by passing in just the name to be inserted as a string, or optionally a placeholder. They will then go on to either run it using an SQLAlchemy engine or by compiling it and running it with the DBMS's API. If they compile it, then it will have a mix of inlined literals and placeholders depending on which of name and num they left as default values.

Is this approach to permitting retention of inlined literal values while compiling BindParameters to placeholders the best way to accomplish my goal with SQLAlchemy? Would you give any other advice for thinking about or solving this problem? I'm pretty new to this system and don't have anyone with more experience to turn to for advice, so I would appreciate any wisdom the community can offer. Thank you!

**Comment by CaselIT** [comment]
Hi,

I would advise against doing that, and instead always use placeholders, since I'm not sure what advantage you achieve by going with your hybrid solution.

Also note that your string handling is not safe if, for example, your string includes a '

**Reply by bskubi** [reply]
The advantage I was hoping to achieve was allowing the user to construct a query for batch execution when some user-specified values are static across all statements to be executed. This is a bioinformatics library, so an example would be if the user wants a chromosome name to be held constant for all statements to be executed, but is inserting a large number (potentially billions) of different scientific measurements for different cells and coordinates on that chromosome.

Since as I mentioned, DuckDB's native execution API is faster than duckdb_engine + SQLAlchemy's execution, I want to construct this statement in SQLAlchemy, compile it, and then execute with DuckDB's native API.

**Reply by CaselIT** [reply]
Can you provide the sql you would like to generate and an example table?
